apis/v1alpha1/httproute_types.go

```diff
@@ -270,6 +270,25 @@ const (
 	HeaderMatchImplementationSpecific HeaderMatchType = "ImplementationSpecific"
 )
 
+// QueryParamMatchType specifies the semantics of how HTTP query parameter
+// values should be compared. Valid QueryParamMatchType values are:
+//
+// * "Exact"
+// * "Presence"
+// * "RegularExpression"
+// * "ImplementationSpecific"
+//
+// +kubebuilder:validation:Enum=Exact;Presence;RegularExpression;ImplementationSpecific
+type QueryParamMatchType string
+
+// QueryParamMatchType constants.
+const (
+	QueryParamMatchExact                  HeaderMatchType = "Exact"
+	QueryParamMatchPresence               HeaderMatchType = "Presence"
```

Comment:
I'm not very familiar with query parameter matching myself, is `Presence` a commonly supported or asked one?

Reply:
https://www.envoyproxy.io/docs/envoy/latest/operations/admin#post--drain_listeners is one example in the wild where a `Presence` could be useful?

It does seem inconsistent with Header at least.

Reply:
To be more clear, I mean there should probably be one common way to describe this in the API. So add Presence to Header or do something else, but they should be consistent.

I think you could express it with regex of "" possibly.

For a prior art, Istio has ~everything as the same StringMatch type: https://istio.io/latest/docs/reference/config/networking/virtual-service/#StringMatch

Reply:
> So add Presence to Header or do something else, but they should be consistent.

But then Presence as PathType makes little sense.
I think it is okay to have different match types for different metadata pieces like query string, header, path, etc because they have different properties and uses.

Reply:
I agree they don't need to be the exact same type, but they should be consistent where it makes sense? Presence in path definitely would be odd. I see headers and query params as pretty similar so I would envision them having the same options

Reply:
> they should be consistent where it makes sense?

We need to dig deeper into what consistency means here.

I agree that they should be consistent for the end-user - kubectl user, meaning having a RegularExpression match on path/header/query param should do the same thing and not surprise the user.
For the implementers, I think having separate Go types and different subsets is okay and good because that makes this API (our go package) easy to use.

Reply:
Yeah that makes sense, I'm fine with dropping Presence from this PR to keep it smaller in scope. We can add it later to both if there's interest.

Reply:
I was only talking about the yaml API surface. Whether we use a single or different go type  I don't have a strong opinion, dedicated types seems reasonable.

Reply:
Yep agree that we should still use separate Go types, kept those in place while dropping "Presence"